typing/types.ml

```diff
@@ -135,36 +135,70 @@ and method_privacy =
   | Mprivate of field_kind
 
 (* Variance *)
+(* Variance forms a product lattice of the following partial orders:
+     0 <= may_pos <= pos
+     0 <= may_weak <= may_neg <= neg
+     0 <= inj
+   Additionally, the following implications are valid
+     pos => inj
+     neg => inj
+   Examples:
+     type 'a t        : may_pos + may_neg + may_weak
+     type 'a t = 'a   : pos
+     type 'a t = 'a -> unit : neg
+     type 'a t = ('a -> unit) -> unit : pos + may_weak
+     type 'a t = A of (('a -> unit) -> unit) : pos
+     type +'a p = ..  : may_pos + inj
+     type +!'a t      : may_pos + inj
+     type -!'a t      : may_neg + inj
+     type 'a t = A    : inj
+ *)
 
 module Variance = struct
   type t = int
   type f = May_pos | May_neg | May_weak | Inj | Pos | Neg | Inv
   let single = function
     | May_pos -> 1
-    | May_neg -> 2
+    | May_neg -> 2 + 4
     | May_weak -> 4
     | Inj -> 8
-    | Pos -> 16
-    | Neg -> 32
-    | Inv -> 64
+    | Pos -> 16 + 8 + 1
+    | Neg -> 32 + 8 + 4 + 2
+    | Inv -> 63
   let union v1 v2 = v1 lor v2
   let inter v1 v2 = v1 land v2
   let subset v1 v2 = (v1 land v2 = v1)
   let eq (v1 : t) v2 = (v1 = v2)
-  let set x b v =
-    if b then v lor single x else  v land (lnot (single x))
+  let set x v = union v (single x)
+  let set_if b x v = if b then set x v else v
   let mem x = subset (single x)
   let null = 0
   let unknown = 7
```

Comment:
Could you define this explicitly as `set May_pos @@ set May_neg @@ null`? This would be clearer and more robust to code changes.

Reply:
Indeed. I just left the original code here, but this would be better that way.